Metric SDK: add asynchronous instrument details; add export pipeline terminology details (#1159)
open-telemetry/opentelemetry-specification: specification/metrics/sdk.md

is defined as the synchronously-copied Aggregator combined with the
Label Set, Resource, and metric Descriptor.

#### Accumulator: Asynchronous instrument state

The set of asynchronous instruments registered through a Meter
implementation each MUST be associated with single, non-null callback
with either a single- or batch-instrument calling signature.

The Accumulator MUST implement [single-value per label set restriction
specified by the API](api.md#asynchronous-calling-conventions) for
asynchronous instruments.  The Accumulator MUST ensure the
last-observed value is the output value, after eliminating duplicate
values.

TODO: _Are there more Accumulator functional requirements?_
- e.g., do callbacks run with timeouts? what should SDKs do if they do not return?

### Processor

TODO _Processor functional requirements_

### Processor checkpoint interval

The Processor checkpoint interval brackets Accumulations into a single

> unit of data.  The Processor checkpoint interval starts and finishes
> before and after calling Accumulator.Collect to process Acumulations

Reviewer: ```suggestion
unit of data.  The Processor checkpoint interval starts before and finishes
after calling Accumulator.Collect to process Acumulations
```

Reviewer: FYI, I'm very interested to see the Processor spec finished. I don't have a good idea of why "start" is necessary on the Processor and how its supposed to react to that call.

Author: This is a good point. Linking these two conversations together here: https://github.com/open-telemetry/opentelemetry-specification/pull/1198#discussion_r527802577

@jkwatson you wrote:

> why not have the controller pass this data to the processor, and decouple the accumulator from the processor?

See also https://github.com/open-telemetry/opentelemetry-go/issues/1362, where this coupling makes trouble for setting up the SDK (FYI @seanschade). Thinking in terms of the OTel-Go implementation, it would make sense for Collect() from the Accumulator to write into a channel and for the Processor to read from a channel.

I can imagine in other languages it would be more natural to use an iterator pattern to iterate over the results of collection using a `ForEach()` pattern, the way exporters consume the output of the `Processor` (although it means executing asynchronous instrument callbacks during the iteration). It won't be easy to do such a refactoring in Go because of how map iteration works--I'll try to make this explanation allow more language-specific approach and encourage decoupling, then file issues to fix the OTel-Go implementation.

### Export pipeline detail

#### Export pipeline data types

**Accumulation**: Passed from the Accumulator to the Processor, this

> consists of the Instrument, the Label Set, the Resource, and a

Reviewer 2: This should only contain a description of a metric instrument, not the instrument itself, right?

```suggestion
consists of the Metric Descriptor, the Label Set, the Resource, and a
```

snapshot of the Aggregator state (i.e., the Aggregation).

**Aggregation**: Output by an Aggregator, this represents state from
one or more metric events.  These have various types (e.g., Sum,
Histogram).

**ExportRecord**: Passed from the Processor to the Exporter, this
consists of the same metadata as the Accumulation, with the
Aggregation corrected for ExportKind, with the associated start and
end timestamps.

**ExportRecordSet**: The set of ExportRecords computed during a
Processor checkpoint interval.

#### Export pipeline configuration

> **AggregationKind**: This is determined by the Aggregation, which is
> returned by an Aggregator, chosen by an AggregatorSelector.
> Aggregation kinds determine what data can be read from an Aggregation
> (i.e., the format).

Reviewer 2: Could we describe these `*Kind` definition more by what they are instead of what determines them and what calls them?

Possibly:

```suggestion
**AggregationKind**: The format which an Aggregator aggregates data into and ultimately sends it as.
```

**ExportKind**: This is determined by the ExportKindSelector and may
be cumulative or delta.  Note that the term ExportKind is used in the
SDK to refer to this choice, while the same concept is called
AggregationTemporality when stored as a field in the OpenTelemetry

> protocol.  TODO: rename ExportKind to AggregationTemporality?

Reviewer 2: I like the idea of consistency :+1:
